projects/hathor-wallet-headless/0004-tx-transaction.md

```diff
@@ -0,0 +1,663 @@
+    token?: string,
+    authority?: `mint` | `melt`,
+    address?: string,
+    checkAddress?: boolean,
+  ): TransactionTemplateBuilder {}
+  addDataOutput(
+    position: number,
+    data: string,
+  ): TransactionTemplateBuilder {}
+  addRawOutput(
+    position: number,
+    amount: number,
+    script: string,
+    token?: string,
+    authority?: `mint` | `melt`,
+  ): TransactionTemplateBuilder {}
+  addShuffleAction(
+    target: 'inputs'|'outputs'|'all',
+  ): TransactionTemplateBuilder {}
+  addFillChangeAction(
+    target: 'inputs'|'outputs'|'all',
+  ): TransactionTemplateBuilder {}
+  build(): TransactionTemplate {}
+  export(): TemplateInstruction[] {}
+}
+```
+
+- `addInstruction`: add a general instruction as an object literal
+- `addUtxoInput`: add the specific UTXO input instruction
+- `addRawInput`: add the specific Raw input instruction
+- `addTokenOutput`: add the specific Token output instruction
+- `addDataOutput`: add the specific Data output instruction
+- `addRawOutput`: add the specific Raw output instruction
+- `addShuffleAction`: add the specific Shuffle action instruction
+- `addFillChangeAction`: add the specific FillChange action instruction
+- `build`: build and returns a `TransactionTemplate` instance
+- `export`: returns an array of instructions added to the builder in order of addition
+
+### Usage example
+
+The following template will mint 0.01 TST, generate another mint authority to the wallet and it will have a 'foobar' data output as the first output.
+The mint deposit inputs are manually added to the transaction and all outputs are shuffled except for the data output at position 0.
+
+```json
+[
+  { "type": "input/utxo", "position": -1, "fill": 2, "token": "00", "autoChange": false },
+  { "type": "input/utxo", "position": -1, "fill": 1, "token": "TST", "authority": "mint", "autoChange": false },
+  { "type": "output/token", "position": -1, "token": "TST", "amount": 1 },
+  { "type": "output/token", "position": -1, "token": "TST", "authority": "mint", "amount": 1 },
+  { "type": "action/change" },
+  { "type": "action/shuffle", "target": "outputs" },
+  { "type": "output/data", "position": 0, "data": "foobar" },
+]
+```
+
+This same transaction template instructions set could be built as:
+
+```ts
+// Token UID for token TST
+let TST = '000cafe...cafe123';
+
+const ttInstructions = new TransactionTemplateBuilder()
+  .addUtxoInput(-1, 2)
+  .addUtxoInput(-1, 2, TST, "mint")
+  .addTokenOutput(-1, 1, TST)
+  .addTokenOutput(-1, 1, TST, "mint")
+  .addShuffleAction("outputs")
+  .addDataOutput("foobar")
+  .export();
+```
+
+We can get a transaction template instance from the instructions set by the `build` or by calling the transaction template `from` with the transaction template instructions:
+
+```ts
+const tt = TransactionTemplate.from(ttInstructions);
+```
+
+Then we can get the designed transaction by calling the `build`:
+
+```ts
+const transaction = tt.build(myWalletFacade);
```

Comment:
question: How complete will be this transaction? Complete enough to be informed at the Explorer "Send Tx" screen, or do we have another reference on how to use this finished tx?

Reply:
This should be an instance of `Transaction` with the inputs, outputs, tokens and some options (version, signalBits, token name and token symbol).

With these the lib will be able to complete (add timestamp, calculate weight) and sign the transaction by using the `SendTransaction` facade like we do with the other facade methods.